flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroSchemaConvertor.java

```diff
@@ -200,4 +241,204 @@ private static DataType handleLogicalTypeSchema(Schema fieldSchema) {
         }
         return null;
     }
+
+    /**
+     * Converts Flink SQL {@link org.apache.flink.table.types.logical.LogicalType} (can be nested)
+     * into an Avro schema.
+     *
+     * <p>Use "org.apache.flink.avro.generated.record" as the type name.
+     *
+     * @param schema the schema type, usually it should be the top level record type, e.g. not a
+     *     nested type
+     * @return Avro's {@link Schema} matching this logical type.
+     */
+    public static Schema convertToSchema(org.apache.flink.table.types.logical.LogicalType schema) {
+        return convertToSchema(schema, "org.apache.flink.avro.generated.record");
+    }
+
+    /**
+     * Converts Flink SQL {@link org.apache.flink.table.types.logical.LogicalType} (can be nested)
+     * into an Avro schema.
+     *
+     * <p>The "{rowName}_" is used as the nested row type name prefix in order to generate the right
+     * schema. Nested record type that only differs with type name is still compatible.
+     *
+     * @param logicalType logical type
+     * @param rowName the record name
+     * @return Avro's {@link Schema} matching this logical type.
+     */
+    public static Schema convertToSchema(
```

Comment:
Two questions:
1. Why not use the same map optimization here as well?
2. Why are we converting Flink's logical type to Avro?

Reply:
We need this for the formation of schema provider we discussed offline. During the instantiation of the sink, we have flink's logical schema available. This could be converted to avro schema without calling the bigquery client again.

Reply:
Fixed.